kad-dht/README.md

```diff
@@ -0,0 +1,427 @@
+# libp2p Kademlia DHT specification
+
+| Lifecycle Stage | Maturity       | Status | Latest Revision |
+|-----------------|----------------|--------|-----------------|
+| 3A              | Recommendation | Active | r0, 2021-05-07  |
+
+Authors: [@raulk], [@jhiesey], [@mxinden]
+
+Interest Group:
+
+[@raulk]: https://github.com/raulk
+[@jhiesey]: https://github.com/jhiesey
+[@mxinden]: https://github.com/mxinden
+
+See the [lifecycle document][lifecycle-spec] for context about maturity level
+and spec status.
+
+[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md
+
+---
+
+## Overview
+
+The Kademlia Distributed Hash Table (DHT) subsystem in libp2p is a DHT
+implementation largely based on the Kademlia [0] whitepaper, augmented with
+notions from S/Kademlia [1], Coral [2] and the [BitTorrent DHT][bittorrent].
+
+This specification assumes the reader has prior knowledge of those systems. So
+rather than explaining DHT mechanics from scratch, we focus on differential
+areas:
+
+1. Specialisations and peculiarities of the libp2p implementation.
+2. Actual wire messages.
+3. Other algorithmic or non-standard behaviours worth pointing out.
+
+For everything else that isn't explicitly stated herein, it is safe to assume
+behaviour similar to Kademlia-based libraries.
+
+Code snippets use a Go-like syntax.
+
+## Definitions
+
+### Replication parameter (`k`)
+
+The amount of replication is governed by the replication parameter `k`. The
+default value for `k` is 20.
+
+### Kademlia routing table
+
+The data structure backing this system is a k-bucket routing table, closely
+following the design outlined in the Kademlia paper [0]. The bucket size is
+equal to the replication paramter `k`, and the maximum bucket count matches the
+size of the SHA256 function, i.e. 256 buckets.
+
+### Alpha concurrency parameter (`α`)
+
+The concurrency of node and value lookups are limited by parameter `α`, with a
+default value of 3. This implies that each lookup process can perform no more
+than 3 inflight requests, at any given time.
+
+### Distance
+
+In all cases, the distance between two keys is `XOR(sha256(key1),
+sha256(key2))`.
+
+## DHT operations
+
+The libp2p Kademlia DHT offers the following types of routing operations:
+
+- **Peer routing** - _Finding_ the closest nodes to a given key (`FIND_NODE`).
+- **Value routing** - _Putting_ a value to the nodes closest to the value's key
```

Comment:
This is not a routing operation (as the section title suggests). Also the description is not accurate.
PUT_VALUE puts a value on the node it is called on. (Not on the "closest node to the key").

Reply:
Changed via 60f4c8d31dda8c4652ef8f6d9b36185f06c72306. Let me know what you think @petar.